activerecord/lib/active_record/transactions.rb

```diff
@@ -350,9 +350,9 @@ def rolledback!(force_restore_state: false, should_run_callbacks: true) #:nodoc:
     end
 
     # Add the record to the current transaction so that the #after_rollback and #after_commit callbacks
-    # can be called.
+    # can be called and state restored.
     def add_to_transaction
-      if has_transactional_callbacks?
+      if has_transactional_callbacks? || has_remembered_start_transaction_state?
```

Comment:
I am not sure how this is solving the issue. 
Why do we need this extra condition here? As far as I can tell, this will not let the record be added to the transaction records array, which is responsible for running the after_commit/rollback callbacks in the end of the transaction. So, if this new condition is true, it won't run callback for that record. Is that the intent of this?